File: 1. 算法部分/第8章 广度优先搜索/2.营救/save.cpp
#include<stdio.h>
#include<string.h>

int BFS(int,int,int,int);

char map[1200][1200];
int qx[3010000],qy[3010000],step[3010000];

int main(void)
{
	int n,m,i,j,ans=0;
	memset(map,'1',sizeof(map));
	scanf("%d",&n);
	m=n;
	getchar();
	for (i=1;i<=n;i++)
	{
		for (j=1;j<=m;j++)
		{
			map[i][j]=getchar();
		}
		getchar();
	}
	int x1,x2,y1,y2;
	scanf("%d%d%d%d",&x1,&y1,&x2,&y2);
	printf("%d\n",BFS(x1,y1,x2,y2));
	return 0;
}


int BFS(int x,int y,int ex,int ey)
{
	if (x==ex&&y==ey) return 0;
	const int gx[4]={0,0,1,-1};
	const int gy[4]={1,-1,0,0};
	int head=1,tail=0,tx,ty,count=0;
	map[x][y]='1';
	qx[head]=x;
	qy[head]=y;
	step[head]=0;
	
	for(;head!=tail;)
	{
		count++;
		tail++;
		tx=qx[tail];
		ty=qy[tail];
		for (int i=0;i<4;i++)
			{
				x=tx+gx[i];
				y=ty+gy[i];
				if (map[x][y]=='1') continue;
				map[x][y]='1';
				head++;
				qx[head]=x;
				qy[head]=y;
				step[head]=step[tail]+1;
				if (x==ex&&y==ey) return step[head];
			}
	}	
	return step[head];
}

File: 1. 算法部分/第8章 广度优先搜索/2.营救/save.pas
program save;
const
  maxn=1000;
  dx:array[1..4]of longint=(1,0,-1,0);
  dy:array[1..4]of longint=(0,1,0,-1);
type
  xy=record
    dis,y,x:longint;
  end;
var
  n:longint;
  g:array[0..maxn+1,0..maxn+1]of boolean;
  dl:array[1..maxn*maxn]of xy;
  sx,sy,gx,gy:longint;
  procedure init;
  var
    i,j:longint;
    ch:char;
  begin
    assign(input,'save.in');reset(input);
    fillchar(g,sizeof(g),false);
    readln(n);
    for i:=1 to n do
    begin
      for j:=1 to n do
      begin
        read(ch);
        g[i,j]:=(ch='0');
      end;
      readln;
    end;
    readln(sy,sx,gy,gx);
    close(input);
  end;

  procedure bfs;
  var
    dir,head,tail,i,j,ti,tj:longint;
  begin
    head:=1;tail:=1;
    dl[head].y:=sy;dl[head].x:=sx;
    dl[head].dis:=0;
    g[sy,sx]:=false;

    repeat
      i:=dl[head].y;j:=dl[head].x;
      for
      dir:=1 to 4 do
      begin
        ti:=i+dy[dir];tj:=j+dx[dir];
        if g[ti,tj] then
        begin
          inc(tail);
          dl[tail].y:=ti;dl[tail].x:=tj;
          dl[tail].dis:=dl[head].dis+1;
          g[ti,tj]:=false;
          if (ti=gy)and(tj=gx) then
          begin
            assign(output,'save.out');rewrite(output);
            writeln(dl[tail].dis);
            close(output);
            halt;
          end;
        end;
      end;
      inc(head);
    until head>tail;
  end;

begin
  init;
  bfs;
end.








